### DISCUSSION TITLE
Champion "Negated-condition if statement"

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8639] [Language Ideas]
- [ ] Proposal added
- [ ] Discussed in LDM (notes appear to be missing)
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See https://github.com/dotnet/csharplang/issues/568 where this idea was recently introduced.

**Comment by iam3yal** [comment]
<s>I support the idea but in my opinion `not` would be much more readable than reusing the negation operator.

Just reusing @HaloFour examples:

```
if not(foo is Bar bar)
{
    // stuff
}

while not(finished)
{
    // stuff
}

do
{
    // stuff
}
while not(finished);
```

And maybe as added benefit:
```
if (foo is not Bar bar)
{
    // stuff
}
```

This can probably be applied to other constructs in the language.</s>.

I'm in favor of `unless` but see [@alrz](https://github.com/dotnet/csharplang/issues/882#issuecomment-327683217) explanation as to why this is more expensive to add.

**Comment by alrz** [comment]
Is this a new statement form or we are going to just remove parentheses from all statements like Swift/Rust?

**Comment by Joe4evr** [comment]
Probably new statement form, Gafter showed in the linked issue that removing the parentheses could cause ambiguity.

**Comment by alrz** [comment]
I see, thanks.

**Comment by RogerWaters** [comment]
I would just replace "!" with "not" because it's much more readable. I currently always write "== false" to clearify that condition is negated. Parentheses should be kept, as they mark begin and end of the condition especialy on multi line conditions.

**Comment by DavidArno** [comment]
My personal opinions:

1. `if (foo is not Bar)` - 👍 👍 
2. `if !(expression)` - 👍 
3. `if not (expression)` - 👎 

I disagree with @eyalsk and @RogerWaters and actually think `!` is more readable than `not` for a simple boolean expression.

**Comment by yaakov-h** [comment]
I'll throw my vote in for `!(expr)` too.

`foo is not Bar` only works for pattern matching, not other expressions where it makes sense to read it one way but you want the condition negated.

`not` is an entirely new keyword that I don't think is necessitated here. We already have `!` throughout the rest of the language, which is well understood. We don't have - or need - `if (foo not == bar)` etc, or `bool a = ...; bool b = not a`, so I don't see the benefit for an `if not`.

**Comment by dsaf** [comment]
I hope this will not be implemented in 7.x-8.x:

1. Why this and not e.g. better LINQ syntax integration? To show that incremental language features can be delivered quickly as part of minor versions?

2. The ```not``` syntax - does it mean interchangeability with ```!```? It's not even in LINQ where it would make the most sense.

3. The ```!``` syntax - will it cause confusion with the coming not-null assertion? Is it that much more readable even? Why not just extract meaningfully named variables and extensions? E.g. I use ```if (x.None())``` instead of ```if (!x.Any())```.

**Comment by iam3yal** [comment]
@yaakov-h 

> foo is not Bar only works for pattern matching, not other expressions where it makes sense to read it one way but you want the condition negated.

Well, that's your opinion which I respect but I disagree, `not` can definitely work here too, it's a matter of taste.

> not is an entirely new keyword that I don't think is necessitated here. We already have ! throughout the rest of the language, which is well understood. 

And this is a new statement, we didn't have `when` before and now we do.

> We don't have - or need - if (foo not == bar) etc, or bool a = ...; bool b = not a, so I don't see the benefit for an if not.

I'm not sure I follow, you wouldn't be able to use `not` inside the if statement, only outside; like I said before we didn't have many things in the language and now we do, in my opinion it's really a weak argument against it, it's more reasonable to say it's not my cup of tea than to say we didn't have it before or try to speak in the name of everyone and say we don't need it, well, again, I prefer `not`. 😄 

I wonder whether `if!(...)` works because some people prefer the following style `if(...)` over `if (...)`.

p.s. Just to make myself crystal clear, I don't mind reusing the negation operator, I just think that `not` would be more readable but than @DavidArno thinks the opposite, fun. 😆

**Comment by HaloFour** [comment]
@eyalsk 

> I wonder whether `if!(...)` works because some people prefer the following style `if(...)` over `if (...)`.

You'd still be able to put a space between `if` and `!(...)` here, e.g. `if !(foo is Bar) { }`.

Personally I think that `!` is plenty readable.  Languages in the C# family generally prefer punctuation for operators rather than keywords and it would be weird to have both `!` and `not` be unary negation operators.  That said I don't think that this proposal obviates the need and use of a `not` pattern and the above could just as easily be written as `if (foo is not Bar)`.  But this proposal is significantly easier to implement and quickly eliminates a niggling aspect of C grammar.

**Comment by MgSam** [comment]
I'm not enthused about this proposal but it's certainly an improvement over today's mess of parenthesis required to do this, `if(!(condition))`.

My main problem with it is that it still forces negative `is` expressions to be inside out. Compute the value of some expression and then negate it, rather than just conceptually checking for the negative expression. This makes it a lot slower to mentally parse what's going on. Thus, I'd still prefer `isnot` (don't care the exact wording) in addition to, or instead of, this proposal.

**Comment by MgSam** [comment]
> Languages in the C# family generally prefer punctuation for operators rather than keywords...

Maybe that was the case at one point in time but not any more. There are a plethora of examples in C# and other languages of word operators. I think readability is much more important than trying to follow design principles from a 40 year old language.

**Comment by HaloFour** [comment]
@MgSam 

> There are a plethora of examples in C# and other languages of word operators.

C# has mostly kept to the same conventions as C/C++.  LINQ and `await` are the big outliers that come to mind and you could argue that `await` falls into the same category as `yield`.  I'm not opposed to `isnot` or `not` but I'd rather view them in a larger context such as pattern matching.  If `not` patterns were closer to being on the table I'd happily prefer that over this proposal.  But this proposal is also infinitely smaller and easier to implement and does not preclude the possibility of including those operators in the future.

From the syntax suggested above, `not` would still require the inside-out evaluation that you dislike about `!`.  How would you suggest supporting that keyword without that being a problem?  Or are you just arguing for the addition of an `isnot` operator which can't be used in general purpose negation scenarios?

**Comment by jnm2** [comment]
@DavidArno 

> My personal opinions:
> 
>  1. if (foo is not Bar) - 👍 👍

What do you mean we never agree on anything? 😆 

(:+1:)

**Comment by MgSam** [comment]
`typeof`, `nameof`, `as`, `is`, `default`, `await`, `yield`, the query comprehension form of LINQ, possibly `with` in the future. Lots of keyword operators. And with `is` getting a major upgrade as part of pattern matching, the trend is only being reinforced.

To make C# pattern matching as usable as possible, it creates the immediate need for a better way to express negation in this context. That's why I'd prefer `isnot` be focused on rather than simply relaxing the parens requirement on `if`, though I'm not opposed to doing both.

My issue isn't about the order of evaluation- it's about how it's mentally parsed when a human being is reading it. I've seen far too many logic errors from people sprinkling `!` operators outside of nested expressions and then it being too hard to mentally parse what's going on so they make a mistake somewhere. And that is only like to get even worse with pattern matching only having a positive form of the expression. This proposal, while it does decrease verbosity, does not solve that part of the problem.

It's directly analogous to the `!=` operator. No language needs a `!=` operator. You could easily express any logic using only the unary `!`. But the huge benefits of being able to directly express a negative condition outweigh the cost of extra syntax.

**Comment by HaloFour** [comment]
@MgSam 

> That's why I'd prefer `isnot` be focused on

Well it's not like the team can't do both.

I'd prefer `not` combined with the `is` operator to keep the syntax consistent regardless of how you use the pattern, e.g. `if (foo is not Bar)` or `case not Bar:`

If it came down to an either-or proposition I would certainly prefer that the team focus on `not`.

>  rather than simply relaxing the parens requirement on `if`

This proposal doesn't go that far.  Actually it's been demonstrated that removing the parens requirement will create syntax ambiguities.  All this proposal does is allow an optional `!` outside of the parenthesis in order to negate the entire condition.

> My main problem with it is that it still forces negative `is` expressions to be inside out.

> My issue isn't about the order of evaluation

Well, which is it?  :grin:

**Comment by alrz** [comment]
>All this proposal does is allow an optional ! outside of the parenthesis in order to negate the entire condition.

 I'd prefer to see it as an `if` followed by an optional `!` i.e. a modifier on `if`: `if! (e)`, rather than an operator on `(e)`: `if !(e)`, which doesn't make sense at all. /s

**Comment by yaakov-h** [comment]
well if `if!(e)` becomes valid (i.e. no space required after `if`), you can enforce that with an analyser... 😄

**Comment by codefox42** [comment]
Sometimes I miss the simplicity of `unless` in [Perl](https://perldoc.perl.org/perlsyn.html#Statement-Modifiers):

    unless (success) { throw new InvalidOperationException("..."); }

**Comment by yaakov-h** [comment]
@codefox42 #138 is probably the closest we'd get.

**Comment by CyrusNajmabadi** [comment]
In a V1 i think i would prefer "not".  However, C# has already shipped 7 (soon to be 8) versions with an existing "not" operator ```!```.  We should use what we have when it makes sense.  We don't need to invent synonyms.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Agreed.  I'd only consider `not` within the context of pattern matching where I believe @alrz demonstrated that reusing `!` would cause ambiguities.  It wouldn't be usable as a replacement to `!`, e.g. `not y == x` would not compile.  But maybe `foo is not Bar` could, as well as `case not 0` and, potentially, `pt is Point(var x, not 0)` whenever recursive patterns get off of the ground.

**Comment by iam3yal** [comment]
@CyrusNajmabadi You and others are probably right about `not`, I thought it was a good idea but anyway, what do you guys think about `unless`? is adding a new keyword for it off the table? just curious.

**Comment by alrz** [comment]
@HaloFour As demonstrated in #277 we should actually consider reusing `!` but I think its precedence would be an issue, and tbh it's not as readable as `not` in a pattern e.g. `!0` doesn't make much sense. 

@eyalsk I think `unless` is relatively more expensive since `unless(e);` is a valid invocation expression and we need to reconstruct it as an `unless` statement with a semicolon in the body after semantic analysis. As long as `if!` works I don't think it would worth it, however I agree it's a little strange compared to other proposed keyword modifiers like `foreach?` and `await?`.

**Comment by iam3yal** [comment]
@alrz I understand, thanks. ;)

**Comment by DavidArno** [comment]
@jnm2,

It's not just you agreeing with me; ~~five~~ ~~six~~ eight people do! Not since [the best bits of pattern matching were dropped from C# 7](https://github.com/dotnet/roslyn/issues/10866) have I found myself being agreed with so much. 😀

**Comment by sirgru** [comment]
With `if` `!` seems more natural, but with `is` `not` seems more natural. If there's both then there's the question why there are 2 ways to do the same thing (code style issues, combination issues). So I would guess `! is` would have to do. I don't like `isnot`, it's like `elseif` and that's just improper English.

Also, I don't like the removing of parens. It makes for sloppy one-liners. It optimizes for typing but not for reading.

**Comment by HaloFour** [comment]
@sirgru 

The issue is that pattern matching extends beyond `is`.  Would we also have `!case`?  What about with recursive patterns where there wouldn't be an additional keyword?  `point is Point(var x, !0)`?

**Comment by DavidArno** [comment]
@HaloFour,

Is the following proposed for recursive patterns?
```cs
point is Point(var x, not 0)
```
If so, I definitely agree that `not` makes a lot more sense here than `!` does and it's a lot more compact than 
```cs
point is Point(var x, var y) when y != 0
```

**Comment by alrz** [comment]
ok I'm convinced if we're going to do `if!` syntax we could do it for all statements and require blocks to disambiguate.

that said, the example @gafter gave becomes a compile-time error.
```cs
if exprDoNotStartWithOpenParen
        (M)(e1)(e2); // ERROR: expected `{`, found `;`
```
then we have no changes in the syntax node, just the parser.

Although, you can't do `if !(e) return;` anymore, instead you should write `if !(e) { return; }`.

A corner case would be object initializers, `if new bool() {}` and `if new bool() {} {}` both may become valid syntax, depending on how we handle the initializer block.

**Comment by sirgru** [comment]
@alrz So we have to have either parens or braces in one liners. What are we saving here by omitting the parens? It introduces a new "unexpected" rule.

**Comment by tannergooding** [comment]
I think the biggest reason for using `not` instead of `!` (or allowing both) is that `!` can be easy to miss sometimes, especially if it is just at the beginning of a complex expression.

Most operator/punctuation characters are either visible by themselves or tend to have whitespace surrounding them to make them more readily visible (most of them, such as `&`, `+`, and `-` are ~10 pixels wide on my screen, others such as `,` and `;` are ~5px). However, `!` is visibly small and is easily swallowed when intermixed with parenthesis and other statements (it is 3px on my monitor, with 2 of those pixels being used for anti-aliasing). -- Note: The pixel sizes are from a 4K screen with 150% scaling (OS scaling, not VS scaling) using the default font in VS.

Allowing `if!(expression)` suffers the same problem as what we have today, the `!` gets swallowed by the surrounding characters and is easy to miss. `if !(expression)` is a bit better (and more readable than today's `if (!(expression))`), but it is still not quite as "parsable" (at least by my eyes) as `if not (expression)` or `if ((expression) == false)`

**Comment by alrz** [comment]
@sirgru 

> So we have to have either parens or braces in one liners

Note that the case for a negated if statement is not conciseness, it's about readability (as mentioned in the linked proposal).


>What are we saving here by omitting the parens?

Rust for instance, did this from day one and nobody ever wanted a "negated if statement". I think that pair of parentheses are the reason we're here. Besides, most of codebases out there are using blocks anyways, because it's more readable and have better diff when modified.

> It introduces a new "unexpected" rule.

I think consistency is important, even though reusing `!(e)` seems like a good idea, it also *is* a valid expression and an embedded-statement right next to it, is not what I'd expect. if we just do this for `if` it *looks* like you are ommiting parens from `if`, but it's actually a new syntax and you don't have it in other statements. *that* is also unexpected.

**Comment by alrz** [comment]
@tannergooding


>However, ! is visibly small and is easily swallowed when intermixed with parenthesis and other statements

Same would apply to the post-fix "not null" operator. The fact that it's post-fix may cause it to be totally overlooked.. so, I think we're stuck with it?

**Comment by tannergooding** [comment]
I don't think it will be quite as bad.

The biggest issue with `!`, today is it tends to not have surrounding whitespace. I think this is why `if (!(expression))` is hard to parse but `if !(expression)` is so much better (even if not "ideal"). `a != b` is likewise, not hard to parse, both due to the surrounding whitespace and due to the height difference between `!` and `=`.

There will certainly be some types where the `not null` operator is hard to read, such as `roof!` (which is also why I probably don't like `if! (expression)` and `if!(expression)`), but it should be a bit more readable Ex: `string! value`, `Exception! value`, `IDisposable! value`, `object! value`, etc.

**Comment by Thaina** [comment]
Agree for `if !(expression)`

And I prefer `obj not Type` than `obj is not Type`. At least it should be `obj isnot Type`

**Comment by jnm2** [comment]
@Thaina what about `case not Type`? Then since we already replace `case` with `obj is`, that composes as `obj is not Type`.

**Comment by Thaina** [comment]
@jnm2 At least `case` is special one that need it own line and special syntax `case SomeThingSomeThing :`. `is` on the other hand tend to stay with many things around it in if clause so `SomeCondition && obj is not Type && SomeOtherCondition` seem a bit hard to read for me

**Comment by jnm2** [comment]
> seem a bit hard to read for me

Meh, I used to be pro `isnot` or `isnt` but I can see it both ways.

**Comment by DavidArno** [comment]
> I would just replace "!" with "not" because it's much more readable...

> [I] think `!` is more readable than `not`...

> I just think that `not` would be more readable...

> ...it's not as readable as...

> ...to make them more readily visible...

I think the one conclusion we can reach here is that a lot of us could probably do with an eye test! 😆

**Comment by alrz** [comment]
I think both `if !()` and `if not` emphasis on the code that would be executed if the condition is not true. But IMO when you want to write `if(!(obj is Foo foo))` the emphasis is on the code that comes *after* the if statement, you just want to make sure that the condition is true on the rest of the block. Maybe we should also consider `guard` statement or something similar here.

So I think these are available options so far.
- A negated if statement: `if!`, `if not` or `unless`.
- A negated pattern, e.g. `if (obj is not T t)`.
- A negated `is` operator, e.g. `if (obj isnot T t)`.
- Allow omission of parens and requiring block.
- A new statement like `guard` with a block that has an unreachable end.
- A new operator like `not` with lower precedence, allowing `if (not obj is T)`.

**Comment by TonyValenti** [comment]
My preference is if!(...)
Also don’t forget while!(...)

I can go either way with “is not” or “isnot”.   That would depend on which allows the IDE team to provide the best experience.

**Comment by alrz** [comment]
I accidentally made myself a negated `if` statement 😆 
```cs
if (condition); else 
{
}
```
This is basically *the* `guard` statement,
```cs
if (argument1 is StatementSyntax statement1); else throw new ArgumentException();
if (argument2 is StatementSyntax statement2); else throw new ArgumentException();
```
Sadly, auto formatter doesn't play nice with it and also you get "possibly mistaken semicolon" warning.

**Comment by alexzzzz** [comment]
My preference is just 
```csharp
if (expression == false)
```
without any extra syntax.

**Comment by gafter** [comment]
@alexzzzz that does not invert the definite assignment state of the expression as `!` does.

**Comment by jordao76** [comment]
I'm in favor of `unless` and `until`. Much more straightforward. I'd even go as far as to say that all the other alternatives are awkward and I'd prefer to live with the status quo.

Now, it's already been stated that they might be harder to implement and that, e.g., `unless(e);` is a valid invocation expression. Although technically correct, I wouldn't rule them out because of these arguably rare occurrences. There might be good solutions to these cases. Also, we should think of the usability of the language overall, regardless of the amount of effort to introduce the feature, unless it becomes prohibitive, which I don't believe to be the case here.

Just my 2 cents...

**Comment by alexzzzz** [comment]
@gafter 
>  that does not invert the definite assignment state of the expression as ! does.

To me it feels much less of a problem than incorrect semantics of `!` most of the time and its poor visibility.

**Comment by CyrusNajmabadi** [comment]
> To me it feels much less of a problem than incorrect semantics of !

What incorrect semantics does ```!``` have?

>  and its poor visibility.

```!``` is very idiomatic C-style language negation.  I don't think visibility of it is a serious problem that needs to be addressed

**Comment by Neme12** [comment]
I used to be rather opposed to `isnot` or even `is not` had it been just an extension of the `is` operator but now seeing that `not` could actually be a pattern that could also be used with `case` and even recursively, that makes a lot of sense to me and it takes away the concern of having a `not` keyword that would just be a replacement for `!`, because `not` as a pattern would be very different and not interchangable - `not 5` and `not Bar` would be a pattern whereas `!Bar` makes not sense at all.

**Comment by HaloFour** [comment]
@Neme12 

I do agree, however the language already supports `!` as a prefix to a pattern, at least in the very specific case of constant `bool` patterns:

```cs
if (b1 is !false) { ... }

switch (b2) {
    case !true: ...
    case !false: ...
}
```

It's more an artifact of the compiler being able to evaluate the Boolean expressions at compile-time so that `!false` is really `true`, but that might set a precedent.  Syntax aside, I think that a `not`/`!` pattern would be quite useful.

**Comment by Neme12** [comment]
@HaloFour 

Yes I realize this is already possible with booleans but I never thought of the `!true` as a negative pattern and I see these as fundamentally different operators.

In my way of thinking this is more like a coincidence, that the negation of `true` is the only value that is `not` true, whereas there is a lot of values that are `not` 5. On the other hand, `!5` makes it seem like it's possible to negate 5 (like you can in C++ and get 0) rather than it being a pattern. It's like the difference between `x == !true` and `x != true`.

And maybe it is a little silly that in this case you could do the same thing in 2 ways, but hey - we already have `x is 5` and `x == 5` and I would argue that this is a much smaller issue/curiosity than even that because this only affects `bool`.

**Comment by Joe4evr** [comment]
> we already have `x is 5` and `x == 5`

I'd say that while these look similar and behave the same, when the compiler breaks them down syntactically, they're actually quite different.
* `x == 5` is simply an expression that translates to `op_Eq(x, 5)`
* `x is 5` gets broken down in a manner of:
  * `x`: local variable
  * `is`: pattern matching keyword
  * `5`: constant pattern

The idea is that many different kinds of patterns could can be used instead of a constant pattern, but patterns can't be used with `==`.

**Comment by Starwer** [comment]
Apparently, the `if(); else` from @alrz made a good laugh...
But I think he has a point: what about allowing the following syntax ?
```C#
if (b1 is Type obj) else 
{
    throw new ArgumentException();
}
```
(Please note: no semicolon)

Even though it sounds awkward if you speak it out, it is unambiguous on what should be expected and enable the Formatter/IntelliSense to sort that out that `obj` is useless and should issue a warning.

**Comment by AustinBryan** [comment]
So in the original post, someone suggested doing `if expr`, instead of `if (expr) which would automatically make `if !(expr)` valid. 

Is the champion for just `if !(expr)` or `if expr` in general?

**Comment by iam3yal** [comment]
@AustinBryan Only `if !(expr)`.

**Comment by AustinBryan** [comment]
@eyalsk Fair. I think `if !(expr)` easier to read than `if (!expr)` and especially `if (!(expr))` which is commonly used

**Comment by CyrusNajmabadi** [comment]
@gafter I'm working on an implementation of this proposal here: https://github.com/dotnet/roslyn/pull/30945 in the features/guard-statements branch.

**Comment by gafter** [comment]
@CyrusNajmabadi : @AlekseyTs is the feature champion.  As such it is his responsibility to shepherd this feature into C# 8.0.  I don't know what his current plans are.

**Comment by CyrusNajmabadi** [comment]
@gafter Thanks!   @AlekseyTs i'd be happy to work with you on this.  I have an implementation approach that i think is viable over at: https://github.com/dotnet/roslyn/pull/30945.  It tries to strike a balance between being correct wrt to the language change, while also minimizing the need for lots of downstream fallout from the language feature.  My hope is that strategy can help lower costs (for both our own implementation, and the impact on the ecosystem) to help make it easier for this to be included in a release sooner.  Let me know what you think!

**Comment by gafter** [comment]
At LDM today we noted that the proposed addition of a negated pattern might make this feature less compelling.  For example, you'd be able to write
``` c#
    if (o is not null) ...
```

**Comment by Thaina** [comment]
@gafter Is it possible to have that negated pattern shorthanded as just `not` ?

```C#
    if (o not null) // contextly obvious to be a pattern
```

**Comment by gafter** [comment]
@Thaina You mean something like an expression that is
> *expression* `not` *pattern*

Sure, that would be possible, but we'd still need the negated pattern forms for `switch`, which means they would work for `is` as well, so this suggested expression form would be quite redundant.

**Comment by CyrusNajmabadi** [comment]
> Sure, that would be possible, but we'd still need the negated pattern forms for switch, which means they would work for is as well, so this suggested expression form would be quite redundant.

@gafter I think i would add more of a rule like this:

```
expr_at_some_precedence_level:
     expr not_pattern

not_pattern:
    not pattern
```

Basically, if there are any 'unary_patterns', then they could be legal directly after `expr` without needing the `is` (as long as they didn't cause ambiguity).

**Comment by gafter** [comment]
@CyrusNajmabadi Right, but you could *also* write

`e is not null`

making the proposed expression form quite redundant.

**Comment by CyrusNajmabadi** [comment]
@gafter  i definitely get that :)   The originating question there was:

> @gafter Is it possible to have that negated pattern shorthanded as just not ?

Basically, the *point* would be just to have the shorthand, even if it was a redundant form.  i.e. will it be so common to write `x is not ...` that it would just be nicer to allow `x not ...`.  I'm personally on hte fence.  It's clean and consistent to not allow and to force 'is'.  But, at the same time, it seems like it could be common enough to be a nice to have.  I don't really care enough to push, but perhaps a read of the LDM would be in order on this?

**Comment by gafter** [comment]
@CyrusNajmabadi I have to say it just reads wrong to me.  `true is not false` is an obviously true sentence, but `true not false` does not even have a verb.

**Comment by marksmeltzer** [comment]
@gafter

`true isnt false`?

I don't think that pattern mixes well with `is` though: `if (a is isnt null)`. The primary pattern reads well enough though: `if (a isnt null)`.

Really, this doesn't come down to rules of English though... This syntax all becomes pure utilitarian convention once we type it 10,000 times. I would be perfectly happen with `if (a not null)`, etc. Using `not` as an operation seems like a useful contraction in specific cases like null comparison. 

In languages like SQL, `not` is most commonly as an adverb though, as you suggest. For example, in SQL's `not in`, the `not` is a modifier to the `in` operation.

I would vote for `not null` being a specifically recognized case. Personally, I would not want to see `not` generalized in the way that @CyrusNajmabadi described above...

**Comment by jnm2** [comment]
I really like `x is not null`.

**Comment by Thaina** [comment]
Well, maybe I just don't like the double separate keyword

`o is not true` might be already good as is. And if we really want to we could just have `isnt` being a shorthand macro that will be transpiled back to `is not`

**Comment by YairHalberstadt** [comment]
@Thaina
C# has no concept of macros, and I don't think this feature would be worth introducing them for...

**Comment by dbmercer** [comment]
Every new syntactic feature we add to the language adds to the cognitive load of a new programmer. It is easy for us, who are experts in C#, to adapt to a little syntactic sugar, but this proposal adds to the syntax of C# without providing a whole lot of value. There is value in consistency of the language, and having as few syntax rules as possible. I voted 👎

**Comment by CyrusNajmabadi** [comment]
@333fred Did we make on a decision on this feature?  My recollection is that we may have explicitly decided to never do it.  But my memory is very fuzzy here.

**Comment by 333fred** [comment]
@CyrusNajmabadi https://github.com/dotnet/csharplang/blob/8889a8eb4ccb04b6ca74277dc567536d321b6137/meetings/2019/LDM-2019-09-11.md?plain=1#L252-L262 is our most recent discussion on it.

**Comment by Thaina** [comment]
I would like to disagree that, after using `is not` it was useful, but it was not overlapped with `negate if`

`is not` is great for negate one condition pattern in declarative manner. But this issue is about negating of the whole `if` condition parentheses. The use case is very difference

**Comment by CyrusNajmabadi** [comment]
On a personal level i went from wanting this a bunch of times to no longer wanting this.  The reason, primarily, was that `!` didn't compose with `is`.  So you had to write `if (!(x is Foo))` which was very ungainly.  The ability for this to be absorbed into a pattern negated (pun intended) the need for the special statement form.

>But this issue is about negating of the whole if condition parentheses. The use case is very difference

Except that for other conditions the `!` can be absorbed into the expression .  e.g. `if (!(a || b))` is easy to write idiomatically as `if (!a && !b)`.  So the presence of `if (!(` has basically gone to almost nothing for me.

**Comment by Thaina** [comment]
`if (!a && !b)` is as ugly as `if (!(a || b))`

The key is the condition are not always short and simple as `a` and `b`. There was a bunch of property and function and there can be more than 4 in one `if` block. The ability to put the negation in front are very obvious. While checking that we need to negate multiple condition is error prone. It not only just `!` it can be `>` `==` `is` and etc

For instance

```C#
if(!myObject0.Value.Condition || !myObject1.Value.CheckCondition() || myObject2.Component is not SomeType || myObject3.GetSomething() == null)
{
}

/// compare with

if !(myObject0.Value.Condition && myObject1.Value.CheckCondition() && myObject2.Component is SomeType && myObject3.GetSomething() != null)
```

This is the real situation

Another common pattern I do is

```C#
if(!(obj is SomeType someObj && someObj != null && someObj?.Value > 0))
    someObj = GetNewObj();
```

**Comment by CyrusNajmabadi** [comment]
> // compare with

They seem equivalently readable to me.  the `!` + `||` version seems clearer as it's how i would say it verbally.  "it's not this or not this or not this etc."

**Comment by ViIvanov** [comment]
> Another common pattern I do is
> 
> ```cs
> if(!(obj is SomeType someObj && someObj != null && someObj?.Value > 0))
>     someObj = GetNewObj();
> ```

It looks like the equivalent code looks much better for me (`obj is SomeType` will never true when `obj is null`):
```cs
if(obj is not SomeType someObj || someObj.Value <= 0)
    someObj = GetNewObj();
```

**Comment by CyrusNajmabadi** [comment]
Or just: `if (obj is not SomeObj { Value: > 0 })`.

**Comment by TahirAhmadov** [comment]
I think this can improve readability by a tiny bit, but it's not worth it. I think VS supports inserting a space after `(` and before `)` in an `if` statement, further reducing the need to make a language change here.

**Comment by Thaina** [comment]
@ViIvanov Sadly not in unity and some framework that will typecast itself to null if it was disposed

**Comment by CyrusNajmabadi** [comment]
That's seems to be a unity issue.

**Comment by Thaina** [comment]
If this could count as issue then I would say allow equal operator to compare with `null` is the bug in language design itself

**Comment by HaloFour** [comment]
@Thaina 

> If this could count as issue then I would say allow equal operator to compare with `null` is the bug in language design itself

Comparing a reference to `null` is idiomatic C# as it was idiomatic C++.  Unity is the only framework that explicitly breaks the model here.

**Comment by Thaina** [comment]
> @Thaina
> 
> > If this could count as issue then I would say allow equal operator to compare with `null` is the bug in language design itself
> 
> Comparing a reference to `null` is idiomatic C# as it was idiomatic C++. Unity is the only framework that explicitly breaks the model here.

I mean overload operator

**Comment by HaloFour** [comment]
@Thaina 

> I mean overload operator

Why would that be a bug with the language?  It's up to the implementation of the operator to apply appropriate semantics, and that would have to handle `null` whether it was a literal or an expression.  IMO it'd be very wrong for the language to have two different behaviors depending on whether you used a `null` literal vs. an expression.

**Comment by Thaina** [comment]
> @Thaina
> 
> > I mean overload operator
> 
> Why would that be a bug with the language? It's up to the implementation of the operator to apply appropriate semantics, and that would have to handle `null` whether it was a literal or an expression. IMO it'd be very wrong for the language to have two different behaviors depending on whether you used a `null` literal vs. an expression.

As for me I don't really count it as a bug. But to allow it then call the one who do it as `issue` or `breaks the model` is not agreeable

I can also say unity is not the only framework that want to do this. To make something become `null` by itself after dispose is actually sensible and convenient. But the best they can do within C# rule is to just fake it with overload operator like this. And it work fine for more than 10 years

Anyway, this is actually nitpicking distraction from the start. I just tell someone that there is this possibility, it not even a main point I have presented, even without that `== null` there can be even more complicate condition. The one who attack unity as if it was their fault is you @CyrusNajmabadi and @HaloFour

**Comment by HaloFour** [comment]
@Thaina 

> I can also say unity is not the only framework that want to do this. To make something become `null` by itself after dispose is actually sensible and convenient. But the best they can do within C# rule is to just fake it with overload operator like this. And it work fine for more than 10 years

It's not fine, which is why folks keep asking Unity and C# to fix it somehow.  It's why Unity objects don't work with reference equality throughout the runtime, not just with C# features.  Breaking the spec is not a good thing, and this is very much Unity's fault.

**Comment by CyrusNajmabadi** [comment]
If you can't compare with null, how would strings implement their == operator?

**Comment by Thaina** [comment]
> folks keep asking Unity and C# to fix it somehow. It's why Unity objects don't work with reference equality throughout the runtime

There are some that was obvious bug from unity side but as far as I can see many want C# to fix itself to allow unity's approach (such as https://github.com/dotnet/csharplang/discussions/2020). It really not as broken as JavaScript

> Breaking the spec

This would be my bad but I never see a spec state that we must not compare with null inside overload operator for equal. People even use overload operator to compare difference class with the same record without care about reference equal

**Comment by HaloFour** [comment]
@Thaina 

> There are some that was obvious bug from unity side but as far as I can see many want C# to fix itself to allow unity's approach (such as https://github.com/dotnet/csharplang/discussions/2020)

Yes, like I said, some folks want C# to change how it works to match Unity's bizarre behavior here.  The team doesn't seem terribly interested given that the feature basically has a single use case, specifically to work around Unity's bizarre behavior.

> This would be my bad but I never see a spec state that we must not compare with null inside overload operator for equal.

Using it to replace reference equality would be abusive at the very least, which is why it has created such a huge amount of friction for Unity developers since it breaks everything that expects reference equality.

**Comment by Thaina** [comment]
> feature basically has a single use case

`using` also has single use case, to defer the dispose of variable. Yet it was very useful. I can say the same for ability to treat null and disposed as the same state. It lessen the amount of code we need to write as null check everywhere

> would be abusive at the very least
> it has created such a huge amount of friction

I can agree with this that unity had abuse C# language in many places but as I said this was allowed from the language and it was worth making framework become so much easy to write a check for state of object. This was one feature that introduce C# to many people to become developer. When object was destroyed it would considerably the same state as `null` is very sensible for common folk

To be fair there was many limitation in C# itself too to make unity require many abusing approach (such as coroutine with yield since the time before async/await) but to call it `breaks the model` is like calling `System.Collections` break the model of C# since we actually should use generic

To be honest I think this feature of unity is very far ahead of C#. The best C# has offer for this situation is `string.IsNullOrEmpty` which is not even an extension method (which means it need to reference the `string` class). While unity ability to check availability of its every object with one same pattern is powerfully simple

Same go with this feature request. When we want to create block of negation of the same logic. Ability to switch between line of `if` to line of `negate if` make us don't need to rewrite the whole line by simply add the negation

Right, this is another use case. Suppose we have write an `if else` block but then we want to swap the block. If we have ability to change just one character in front of the block in addition to dragging the whole block up and down will be very convenient
